DBAPI: Add param to control capturing of db.statement.parameters

stschenk opened the pull request:
# Description

Switched DBAPI so that it will not capture parameterized query parameters.  Capturing this would be a problem in production environments since they will contain sensitive information such as session tokens and hashed passwords.

Have added a parameter to the instrumentation constructor to allow for this feature to be turned on.

I will mark this as a breaking change since I am removing the db.statement.parameter span attribute and something might be depending on it. 

Fixes # (issue)
https://github.com/open-telemetry/opentelemetry-python/issues/1215

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Unit Tests

# Checklist:

- [ x] Followed the style guidelines of this project
- [ x] Changelogs have been updated
- [ x] Unit tests have been added
- [ ] Documentation has been updated

NathanielRN commented:
Hello! I have a PR to move some files you have in this PR to the Contrib repo, please let me know if this gets merged before the PR in the Contrib repo. Please see https://github.com/open-telemetry/opentelemetry-python-contrib/pulls/87

lzchen commented on instrumentation/opentelemetry-instrumentation-dbapi/tests/test_dbapi_integration.py:
```diff
@@ -53,6 +53,50 @@ def test_span_succeeded(self):
+        span = spans_list[0]
+        self.assertEqual(span.name, "testcomponent.testdatabase")
+        self.assertIs(span.kind, trace_api.SpanKind.CLIENT)
+
```
Shouldn't there be a check that the param is in the span attribute?

lzchen commented on instrumentation/opentelemetry-instrumentation-dbapi/tests/test_dbapi_integration.py:
```diff
@@ -53,6 +53,50 @@ def test_span_succeeded(self):
+        self.assertIs(
+            span.status.canonical_code,
+            trace_api.status.StatusCanonicalCode.OK,
+        )
```
Should there be an explicit check that the param span attribute does not exist?

lzchen commented on instrumentation/opentelemetry-instrumentation-aiopg/tests/test_aiopg_integration.py:
```diff
@@ -199,7 +199,11 @@ def test_span_succeeded(self):
+            connection_attributes,
+            capture_parameters=True,
         )
         mock_connection = async_call(
```
Explicitly check the span attribute existence in this test as well.

lzchen commented on instrumentation/opentelemetry-instrumentation-aiopg/tests/test_aiopg_integration.py:
```diff
@@ -199,7 +199,11 @@ def test_span_succeeded(self):
         }
         db_integration = AiopgIntegration(
-            self.tracer, "testcomponent", "testtype", connection_attributes
+            self.tracer,
```
Aren't there a lot of other instrumentations that use dbapi as well (mysql, psycopg2, etc)? I think tests would need to be added for those as well because this is a breaking change.

stschenk replied:
I would argue that since the the other instrumentations already are not testing for "db.statement.parameters", we would not need to modify the test to do so now.

Also db.statement.parameters is not even part of the specification: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md#call-level-attributes, why are we collecting it at all?

lzchen replied:
> I would argue that since the the other instrumentations already are not testing for "db.statement.parameters", we would not need to modify the test to do so now.

To ensure completeness of the feature, tests should be added for each instrumentation. The original authors should have tested for this, but since your change will effect the behavior of these instrumentations, it would be good to have them there as well.

> why is it being collected at all?

That is a good question. Perhaps this should be the discussion topic to address first? @codeboten

lzchen commented on instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py:
```diff
@@ -211,6 +211,7 @@ def __init__(
         connection_attributes=None,
         version: str = "",
         tracer_provider: typing.Optional[TracerProvider] = None,
+        capture_parameters: bool = False,
```
How can this be configured by the user? As well as the instrumentations that depend on this?

stschenk replied:
Yeah, I did not make this configurable for the user.  I initially though I was going to but then came across this PR that removes the capturing of "db.statement.parameters" for AsyncPGInstrumentor
https://github.com/open-telemetry/opentelemetry-python/pull/854 Since the approach was accepted I thought it would be fine for DBAPI too.

Would it be better to not have the constructor param and instead introduce a user settable property? 

Thanks!

lzchen replied:
Having users be able to configure it is fine. The reason why the `AsyncPGInstumentor` works is because the configuration property is exposed to the users (because they call `AsyncPGInstrumentor` constructor to instrument`). However for this case it is in `DatabaseApiIntegration`, which users do not interact with (for dbapi, they use `trace_integration(...)` or for something like mysql, they use MySQLInstrumentor()`.

lzchen reviewed (changes requested):
Some comments.

stschenk commented:
@lzchen / @owais - I have put together a different more conservative and backward compatible PR that will allow for the user to turn off the collection of db.statement.parameters through the use of an env var: https://github.com/open-telemetry/opentelemetry-python/pull/1306.  If this one seems fine, I will go ahead and abandon this PR.

lzchen commented:
@stschenk 
How come we can't go with the same mechanism that `AsyncPGInstrumentor` is using? With the env vars, now we have two different ways of setting the same feature, which is confusing for the user.

stschenk commented:
@lzchen - Good question...  I misinterpreted your previous comment around needing to have the ability for the user to configure the capture of db.statement.parameters as meaning there should be an env prop.

I have changed this PR a bit by doing the following:
- Added the ability to configure the feature through the trace_integration method.  Making the capture off by default for DBAPI
- Left the other DB instrumentations as they are...  They will continue to capture db.statement.parameters.

lzchen commented on instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py:
```diff
@@ -127,6 +133,7 @@ def wrap_connect_(
+            capture_parameters=capture_parameters,
         )
         return db_integration.wrapped_connection(wrapped, args, kwargs)
 
```
I believe `instrument_connection` needs to have this parameter as well.

stschenk replied:
makes sense.

lzchen commented on instrumentation/opentelemetry-instrumentation-dbapi/src/opentelemetry/instrumentation/dbapi/__init__.py:
```diff
@@ -98,6 +101,7 @@ def wrap_connect(
     connection_attributes: typing.Dict = None,
     version: str = "",
     tracer_provider: typing.Optional[TracerProvider] = None,
+    capture_parameters: bool = True,
```
The default should be `False` for all instrumentations that use `dbapi`.

stschenk replied:
ok

lzchen commented:
@stschenk 
We are moving all of our instrumentation code into the [contrib](https://github.com/open-telemetry/opentelemetry-python-contrib) repo. Do you mind opening up this PR there and I can close this one?

stschenk commented:
Sure, will do it first thing tomorrow

stschenk commented:
This has been moved to https://github.com/open-telemetry/opentelemetry-python-contrib/pull/156.

I will go ahead and close this PR
